src/runtime/write_debug_image.cpp

```diff
@@ -108,16 +108,19 @@ WEAK bool ends_with(const char *filename, const char *suffix) {
 }
 
 struct ScopedFile {
-    void *f;
-    ALWAYS_INLINE ScopedFile(const char *filename, const char *mode) {
-        f = fopen(filename, mode);
+    void *const user_context;
+    void *const f;
+    ALWAYS_INLINE ScopedFile(void *user_context, const char *filename, const char *mode)
+        : user_context(user_context), f(fopen(filename, mode)) {
+        halide_msan_annotate_memory_is_initialized(user_context, &f, sizeof(f));
     }
     ALWAYS_INLINE ~ScopedFile() {
         if (f) {
             fclose(f);
         }
     }
     ALWAYS_INLINE bool write(const void *ptr, size_t bytes) {
+        halide_msan_annotate_memory_is_initialized(user_context, ptr, bytes);
```

Comment:
Is this one necessary? We should annotate at the call site of this function when needed, not here. Writing uninitialized memory to a file does seem like it should in fact be an msan failure.

Reply:
The problem is that the memory isn't uninitialized, but msan's interceptor for fwrite() thinks it is:
- Calling app allocates a large stack buffer, only uses a small amount of it, so msan adds a 'shadow' record noting that most of this stack area is uninitialized
- Calling app later calls into Halide and calls this function
- This function always copies from heap memory into a 4k stack buffer for I/O purposes (so that it can ignore variations in strides)
- If we get (un)lucky, that stack buffer overlaps with the area of the buffer from step #1; since our runtime isn't compiled with MSAN enabled, the 'shadow' records aren't updated
- Call into fwrite(), which is intercepted by msan, sees that there is a shadow record for this memory range, sees it has uninitialized bytes, and goes boom

That said... this is a gigantic hack and I probably should rethink this; it assumes/relies on the fact that everything we pass to fwrite() goes through a stack allocation of some sort (which is true) but doesn't actually enforce it (so code changes could easily break this).

Alternate approaches that might make sense:
- Move all buffering for `fwrite()` into the ScopedFile class, so we can have ironclad assurance that the memory is something safe to mark properly. (Less hacky code, but would still let us write potentially-random bytes to file)
- Do msan-specific compiles for *all* runtime modules for msan, so that our runtime code isn't a special flower. (Downside is that our embedded bitcode ~doubles in size for a rarely-used case, and potentially dramatically slows down runtime code in this case too, though expecting decent performance in an MSAN run is not a thing in the first place)
- Just declare that this function can't be used in conjunction with MSAN (and add some checks to ensure that the failure is reliable)

Reply:
Maybe a simple solution would be to use `__msan_check_mem_is_initialized` to check the memory before we buffer it, and then annotate it as initialized after we buffer it?